### PR TITLE
Add exp299 delta-stream contact experiment

### PR DESCRIPTION
**Pull request by znichols**
## Summary
- Set up experiment #299 for compact delta-stream contact documents.
- Replace the invalid V1 interleaved document with V2 sequence-prefix documents:
  `DOC_START AA... CONTACTS_BEGIN DELTA* STOP ... DOC_END`.
- Remove the ad-hoc pairwise-logprob scorers and their withdrawn R-precision table.
- Retain only the packed-cache and distributed-training path, configured for V2 artifacts.
- Document that future evaluation must produce rollout vote matrices and use the unchanged exp82/exp89 reference metric path.

## Why V2

V1 predicted early-residue contacts before seeing future amino acids. V2 supplies the entire amino-acid sequence before any contact suffix, matching the conditioning structure of contacts-v1.

## Validation
- `cd experiments/exp299_models_delta_stream_contacts && uv run --with pytest pytest test_delta_stream_documents.py -q`
- `cd experiments/exp299_models_delta_stream_contacts && uv run python build_summary.py`
- `cd experiments/exp299_models_delta_stream_contacts && uv run python -m py_compile compute_contacts_delta_stream_documents.py build_packed_delta_stream_cache.py dispatch_delta_stream_full.py`

### CODE REVIEW

**Comment by znichols** [comment]
Addressed Codex feedback in 5cd801a5:

- validation shards are now excluded from train in both cache builders
- R-precision comparison intersects available proteins across all score dirs before summarizing
- delta-stream and contacts-v1 scorers now emit per-input timings CSVs
- added local `marinfold-models` path dependency and refreshed `uv.lock`
- added `build_summary.py`, `summary_narrative.md`, preliminary R-precision plot, and `plots/summary.pdf`

**Comment by znichols** [comment]
## Summary
- Set up experiment #299 for compact delta-stream contact documents.
- Replace the invalid V1 interleaved document with V2 sequence-prefix documents:
  `DOC_START AA... CONTACTS_BEGIN DELTA* STOP ... DOC_END`.
- Remove the ad-hoc pairwise-logprob scorers and their withdrawn R-precision table.
- Retain only the packed-cache and distributed-training path, configured for V2 artifacts.
- Document that future evaluation must produce rollout vote matrices and use the unchanged exp82/exp89 reference metric path.

## Why V2

V1 predicted early-residue contacts before seeing future amino acids. V2 supplies the entire amino-acid sequence before any contact suffix, matching the conditioning structure of contacts-v1.

## Validation
- `cd experiments/exp299_models_delta_stream_contacts && uv run --with pytest pytest test_delta_stream_documents.py -q`
- `cd experiments/exp299_models_delta_stream_contacts && uv run python build_summary.py`
- `cd experiments/exp299_models_delta_stream_contacts && uv run python -m py_compile compute_contacts_delta_stream_documents.py build_packed_delta_stream_cache.py dispatch_delta_stream_full.py`

**Comment by znichols** [comment]
## Mature V2 run completed and final checkpoints evaluated

The fresh `full_78500` run completed successfully at step 78,499. I exported, mirrored, and evaluated steps 66k, 72k, 78k, and final 78,499 on all 670 units with 100 rollouts/protein, and refreshed both notebook curves.

### Final matched result

| model | CV1-equivalent step | legacy R/all | legacy R/long | legacy AUC/all | legacy AUC/long |
|---|---:|---:|---:|---:|---:|
| V2 final 78,499 | 71,363 | **0.5763** | **0.5160** | 0.9245 | 0.9028 |
| exp177 final 71,359 | 71,359 | 0.5113 | 0.4595 | 0.9246 | 0.9033 |

V2-minus-exp177 paired differences:

- R/all: **+0.0651**, 95% CI `[0.0547, 0.0757]`
- R/long: **+0.0565**, 95% CI `[0.0450, 0.0684]`
- AUC/all: -0.00005, CI `[-0.0046, 0.0046]`
- AUC/long: -0.00055, CI `[-0.0065, 0.0054]`

Thus V2 finishes with a resolved R-precision advantage and statistically tied AUC at essentially exact matched protein exposure.

Final V2 `eval-val` R/all / R/long: **0.5147 / 0.4842**. Final `eval-denovo`: **0.5494 / 0.4779**.

### Training synopsis

- 78,499 steps; 82.31B packed tokens
- ~94.73M protein-document exposures; 1.42 packed-cache passes
- 112.9 wall-clock hours on 2 nodes × 4 GB200
- Mean throughput 205,396 tokens/s (5.105 s/step)
- Final train loss 0.9190
- Final and best native validation loss 0.91425
- LR ended at 1e-4 after peaking at 1e-3

Notebook-ready `legacy554_*` and `eval_denovo_*` CSVs now include the complete mature curve through final step 78,499.

**Comment by znichols** [comment]
## Exp277 full-corpus V2 conversion + training smoke complete

Converted all four exp277 first-epoch contacts-v1 token caches directly to canonical delta-stream V2 in-region, with strict source parsing and V2 round-trip validation:

| corpus | documents | source shards |
|---|---:|---:|
| native AFDB | 3,963,003 | 125 |
| native ESM | 65,553,178 | 1,669 |
| ProteinMPNN AFDB | 31,702,680 | 199 |
| ProteinMPNN ESM | 130,872,044 | 3,338 |
| **total** | **232,090,905** | **5,331** |

Output row counts exactly match all source document counts. Converted documents:

`s3://marin-us-east-02a/protein-structure/MarinFold/exp299_contacts_delta_stream_v2_sequence_prefix/exp277_full_epoch_documents/2026.09.22.1/`

The append-only vocabulary is now 4,030 IDs, preserving every original exp299 ID and adding signed deltas ±1,025…±1,999. A long-chain tail smoke exercised the extension: 187/1,000 documents used extended deltas, with max sequence length 1,263, max document length 5,627 tokens, and max absolute delta 1,255.

A subsequent cache/training smoke read 1,000 converted ProteinMPNN-AFDB documents (4,812,542 raw tokens), packed them to 8,192 tokens while preserving segment boundaries and blocking cross-document attention, and completed 10 optimizer updates of the exp277 1.5B architecture on 8 H100s:

- W&B: https://wandb.ai/open-athena/MarinFold/runs/delta-v2-exp277-full-epoch-1_5b-smoke-a03
- Iris: `/zack/exp299-v2-exp277-train-smoke-driver-a03`
- Steady update duration: 10.32 s
- Throughput: 101,603 tokens/s
- Final train loss: 4.8388

This verifies conversion, extended token IDs, tokenizer/model vocabulary, online packing, blocked attention, forward/backward, and checkpointing. The independent validation conversion, complete token/length census, and finite four-corpus packed-example count remain before deriving the exact one-epoch update count or launching full training.

**Comment by znichols** [comment]
## Exp277 V2 data preparation and context census complete

Converted the independent validation cache and completed a full vectorized census of every converted training and validation document.

### Context-window result

- Training documents: **232,090,905**
- Training V2 tokens: **202,179,778,565**
- Validation documents: **41,954**
- Validation V2 tokens: **39,106,868**
- Documents over 8,192 tokens: **0**
- Documents exactly 8,192 tokens: **0**
- Longest training document: **6,791 tokens** (1,401 tokens of headroom)
- Longest validation document: **5,859 tokens**
- Training documents over 6,144 tokens: **3,352**
- Training documents over 7,168 tokens: **0**

No truncation, filtering, or context-length increase is needed. Conversion itself is also fail-loud above 8,192, so an over-context document could not have silently entered the output.

The full training corpus is 202.18B V2 tokens versus 248.58B source contacts-v1 tokens. The extended delta range is used by 29,420 training documents and 40 validation documents.

### Prepared artifacts

- Converted documents:
  `s3://marin-us-east-02a/protein-structure/MarinFold/exp299_contacts_delta_stream_v2_sequence_prefix/exp277_full_epoch_documents/2026.09.22.1/`
- Full census:
  `s3://marin-us-east-02a/protein-structure/MarinFold/exp299_contacts_delta_stream_v2_sequence_prefix/exp277_full_epoch_census/2026.09.22.1/`
- Finite variable-length Levanter caches (four training corpora plus independent validation):
  `s3://marin-us-east-02a/protein-structure/MarinFold/exp299_contacts_delta_stream_v2_sequence_prefix/exp277_full_epoch_tokenized_cache/2026.09.22.1/`
- Notebook-ready census:
  `experiments/exp299_models_delta_stream_contacts/data/exp277_v2_document_census.csv`

All five cache builds succeeded, and every cache's element/token totals exactly match the converted parquet census. Next is only the finite packing audit and exact one-epoch update count; no additional data conversion is outstanding.

**Comment by znichols** [comment]
### Finite packing audit now complete

Following the document/cache census above, I reproduced Levanter's exact contiguous greedy packing recurrence (8,192 tokens, at most 64 documents/example) over each corpus in cache order.

| corpus | packed examples | padding |
|---|---:|---:|
| native AFDB | 490,852 | 9.81% |
| native ESM | 7,541,763 | 7.99% |
| ProteinMPNN AFDB | 3,971,561 | 11.05% |
| ProteinMPNN ESM | 14,938,761 | 7.85% |
| **training total** | **26,942,937** | **8.40%** |
| validation | 5,302 | 9.96% |

Training therefore contains **220,716,539,904 packed tokens** and requires **210,492 optimizer updates** at global batch 128 for one finite pass (final batch partial). This replaces exp277 contacts-v1's 266,345-update count; we are not copying that count across formats.

As a cross-check, directly constructing Levanter's `GreedyPrepackedDataset` over the completed validation cache independently returned exactly 5,302 examples.

Notebook-ready details: `experiments/exp299_models_delta_stream_contacts/data/exp277_v2_packing_audit.csv`

At this point document conversion, independent validation conversion, context audit, variable-length cache construction, and finite packing accounting are complete. The remaining work is training-configuration adaptation and cost projection, not data preparation.

**Comment by znichols** [comment]
## Full finite-epoch V2 run configured and submitted

The production config now uses the audited finite corpus directly:

- 26,942,937 packed examples / 210,492 updates at global batch 128
- 8,192-token context with blocked cross-document attention
- independent 5,302-example validation set
- exp277 1.5B Qwen architecture
- WSD optimizer: LR 1e-3, WD 0.2, 10% warmup, 70% stable, 20% linear decay to 1e-4
- scratch initialization, model/data seed 0
- permanent native checkpoints at approximately 10% intervals
- no automatic HF export

The finite loader concatenates the four packed source datasets, shuffles once, verifies the exact expected example count, and does not cycle the mixture after a source is exhausted.

I first submitted the requested 4×4 GB200 gang:

`/zack/exp299-v2-exp277-full-epoch-driver-4x4gb200-a01`

After approximately 13 minutes all four tasks remained `SchedulingGated` with no allocation, so I cancelled it and submitted the requested 2×4 fallback:

`/zack/exp299-v2-exp277-full-epoch-driver-2x4gb200-a02`

Child job:

`/zack/exp299-v2-exp277-full-epoch-driver-2x4gb200-a02/delta-v2-exp277-full-epoch-1_5b-2x4gb200-a02`

The 2×4 job is currently queued with both tasks `SchedulingGated`; no W&B run exists yet because no worker has started. I am leaving this request queued rather than reducing below the requested fallback size.

**Comment by znichols** [comment]
## Corrected 08a launch is running

The mistakenly targeted 02a request was cancelled before allocation. With explicit approval, I mirrored the prepared cache to the GB200 region:

- Source: `s3://marin-us-east-02a/.../exp277_full_epoch_tokenized_cache/2026.09.22.1`
- Destination: `s3://rhoarnet-us-east-08a/.../exp277_full_epoch_tokenized_cache/2026.09.22.1`
- Verified: **32,146 objects / 159,232,199,956 bytes**
- Mirror job: `/zack/exp299-v2-exp277-mirror-cache-08a-a01`

The requested 4×4 GB200 gang placed immediately in `cw-us-east-08a` and is training:

- Driver: `/zack/exp299-v2-exp277-full-epoch-driver-4x4gb200-08a-a03`
- Worker: `/zack/exp299-v2-exp277-full-epoch-driver-4x4gb200-08a-a03/delta-v2-exp277-full-epoch-1_5b-4x4gb200-08a-a03`
- W&B: https://wandb.ai/open-athena/MarinFold/runs/delta-v2-exp277-full-epoch-1_5b-4x4gb200-08a-a03

All four workers are running. Step 0 completed successfully after compilation. Through step 89, steady-state median duration was **2.5805 s/update** (~406k tokens/s), projecting roughly **6.3 days** for 210,492 updates before checkpoint/evaluation overhead. Loss had fallen to 3.81.
